remove Consumer type from Cilium

ianvernon opened the pull request:
The Consumer object did not possess any useful state, as a variety of its
fields were not in use throughout the policy package. These fields included
a Decision, which was never populated during policy calculation, and Reverse,
a field which was never populated at all.
In a Consumable, there was a mapping of security identities from which ingress
traffic is allowed to Consumers. But, once the aforementioned unused fields were
removed, the only useful information within the Consumer was the identity of
the Consumer, which was key in the aforementioned map within a Consumable, and
the DeletionMark within a Consumer, which indicated whether the given entry
in the map should be garbage collected after policy recalculation for a
Consumable. Given that the key in the Consumable's map was  already the same
security identity as that within the Consumer, that left the DeletionMark as
the only field within a Consumer that was necessary to keep. Thus, it made
sense to get rid of the notion of a Consumer, and just have the map within the
Consumable map from a security identity to a boolean indicating whether it
should be garbage-collected or not during policy recalculation.
    
This patch removes the notion of a Consumer from the policy package, and defined
the fields within a Consumable to refer to ingress security-identity-based
policy. This sets the groundwork for adding identity-based egress policy
within Cilium, and  makes the code more readable and easier to understand.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: change key of Consumers from string to NumericIdentity

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: remove GetConsumables function

This function is never used and thus is dead code, so remove it.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: remove Decision field from Consumer

This field never actually stored any useful information about policy decision.
It was never set to anything other than api.Allowed. As such, remove it.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: added ReportFailed for conntrack test

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: remove Reserved field from Consumer

This field is useless, as it is never populated.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: get rid of functions with Consumer receiver

StringID isn't use anymore anywhere, and NewConsumer doesn't provide much value.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
change Consumable Consumers map to map from identity to boolean

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
change Consumable ReverseRules to map to boolean instead of Consumer

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: get rid of Consumer type

Signed-off by: Ian Vernon <ian@cilium.io>:

Commit:
pkg/maps/policymap: get rid of Consumer in function names

Change functions to refer to identities instead of Consumers.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: rename Consumers to Ingress Identities

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
change code to not contain references to Consumer

Change variable names, code comments to remove notion of Consumer.

Signed-off by: Ian Vernon <ian@cilium.io.>

Commit:
api: change allowed-consumers to allowed-ingress-identities

Reflect removal of Consumers in API.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg: update API models to reflect removal of Consumer

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: update test to reflect change in API

Reflect change in JSON for endpoint policy from allowed-consumers to allowed-ingress-identities.

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme commented on pkg/policy/consumer.go:
```diff
@@ -216,40 +200,43 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
+// IngressIdentities map, and thus had to be added, false if it is already added.
+func (c *Consumable) AllowIngressIdentityLocked(cache *ConsumableCache, id NumericIdentity) bool {
+	isIdentityAllowed := c.isIdentityAllowed(id)
+	if isIdentityAllowed == false {
```
We still need to differentiate between the cases of an map entry not existing before and one existing with a `false` value. The reason is that all identities are marked as `false` for garbage collection purposes, but we want to return `true` from this function only if a new entry was created, and `false` if we changed the `false` to `true` meaning we are keeping an existing entry. We also do not want to log about a new ingress identity or call `c.addToMaps` for identities that were already allowed.

```
isIdentityAllowed, exists := c.isIdentityAllowed(id)
if !exists {
```

This requires to change `isIdentityAllowed` to return the second boolean return value from the map lookup.

ianvernon replied:
I'll just get rid of the function and access the map directly in that case.

ianvernon replied:
Fixed.

jrajahalme commented on pkg/policy/consumer.go:
```diff
@@ -216,40 +200,43 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
 		return true
 	}
-	consumer.DeletionMark = false
+
```
Need to `c.IngressIdentities[id] = true` here as well to mark an existing entry to not be garbage collected.

ianvernon replied:
Good catch.

jrajahalme commented on pkg/policy/consumer.go:
```diff
@@ -258,37 +245,34 @@ func (c *Consumable) AllowConsumerAndReverseLocked(cache *ConsumableCache, id Nu
-
-		if consumer.Reverse != nil {
-			c.deleteReverseRule(id, c.ID)
-		}
```
How is the removal of the reverse rule done now?

ianvernon replied:
The Consumable has a separate map, `ReverseRules`.

jrajahalme replied:
But now we don't delete the (reverse) rule from that map like we did here, by calling `c.deleteReverseRule()`. We still add entries to that map like before (in `AllowConsumerAndReverseLocked()`) , though. IMO we should still be able to track if a reverse entry was added and then remove it here correspondingly.

ianvernon replied:
This code was never invoked. The `Reverse` field of any Consumer was never populated with anything (i.e., it was always `nil`), and thus this codepath was never reached. That was part of my rationale for removing the Consumer type all-together, because most of its fields were not used (Decision and Reverse) or Redundant (Identity).

ianvernon replied:
The function also specifies that the given id should be removed from the IngressIdentities map for this Consumable (i.e., ingress is no longer allowed from id to this Consumable).

Are you saying that we should invoke a call to `deleteReverseRule` for this Consumable  anyway? That would perform a lookup in the ConsumableCache for the Consumable corresponding to `id`, and then if that Consumable's `ReverseRules` contained `c.ID`, we would delete it as well. If not, then it's a no-op. Thus, we don't really need a check to call `c.deleteReverseRule(id, c.ID)`, I guess?

aanm replied:
>This code was never invoked.

which code? the reverse map was created here:

```
// allowConsumer must be called with global endpoint.Mutex held
func (e *Endpoint) allowConsumer(owner Owner, id policy.NumericIdentity) bool {
	cache := policy.GetConsumableCache()
	if !e.Opts.IsEnabled(OptionConntrack) {
		return e.Consumable.AllowConsumerAndReverseLocked(cache, id)
	}
	return e.Consumable.AllowConsumerLocked(cache, id)
}
```

ianvernon replied:
@aanm by 'this code' I meant the call to `deleteReverseRule` within the if-condition:
```
if consumer.Reverse != nil {
	c.deleteReverseRule(id, c.ID)
}
```

What I meant was that the `Reverse` field of a Consumer was never non-nil.

jrajahalme replied:
OK, so even when an entry was added to ReverseRules, the `Reverse` field of a consumer was never set to indicate it? So this has never worked as intended, meaning that the `BanConsumerLocked()` never reversed the changes of `AllowConsumerAndReverseLocked()`? If so you should at least leave a TODO etc. to indicate that.

ianvernon replied:
OK, will add a TODO.

jrajahalme commented on pkg/policy/consumer.go:
```diff
@@ -258,37 +245,34 @@ func (c *Consumable) AllowConsumerAndReverseLocked(cache *ConsumableCache, id Nu
 	c.Mutex.RUnlock()
-	return consumer != nil && consumer.Decision == api.Allowed
+	return identityAllowed != false
 }
```
Would be cleaner to defer the RUnlock and then `return c.isIdentityAllowed(id)`

ianvernon replied:
Done.

aanm replied:
Defer adds overhead in small functions. Is is ok if it goes like this?
```golang
func (c *Consumable) Allows(id NumericIdentity) (isAllowed bool) {
 	c.Mutex.RLock()
 	isAllowed = c.getConsumer(id) != false
 	c.Mutex.RUnlock()
        return
}
```

ianvernon replied:
> Defer adds overhead in small functions. Is is ok if it goes like this?

I have no strong opinion on this, so I'll revert that change.

ianvernon replied:
Fixed in latest upload.

jrajahalme reviewed (dismissed):
Seems like this PR should be a pure refactoring effort without any functional or observable behavior changes. However I may have spotted two places that now behave differently.

Commit:
pkg/policy: remove isIdentityAllowed function

Also fix logic for AllowsIngressIdentityLocked to correctly determine whether the
provided identity is in the map or not.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/policy/consumer.go:
```diff
@@ -198,14 +177,14 @@ func (c *Consumable) addToMaps(id NumericIdentity) {
 
 func (c *Consumable) wasLastRule(id NumericIdentity) bool {
-	return c.ReverseRules[id] == nil && c.Consumers[id.StringID()] == nil
+	return c.ReverseRules[id] == false && c.IngressIdentities[id] == false
```
I think I'll need to make a change similar to the one proposed here: https://github.com/cilium/cilium/pull/2769#discussion_r167647012 to check for existence rather than the zero-value of the map since the entry itself can be set to false for GC. Will fix in an upcoming commit.

Commit:
pkg/policy: fix logic in wasLastRule

Checking if the value in the map is false is not enough; we need to check if
the entry exists to determine if we can remove the identity from the policy
maps for a Consumable.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: add TODO for ReverseRules clean up

Add link to #2795, which documents issues with the implementation of ReverseRules, which
is populated when ConnTrack is not enabled for an endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/policy/consumer.go:
```diff
@@ -258,37 +246,36 @@ func (c *Consumable) AllowConsumerAndReverseLocked(cache *ConsumableCache, id Nu
+// Must be called with the Consumable mutex locked.
+func (c *Consumable) RemoveIngressIdentityLocked(id NumericIdentity) {
+	if _, ok := c.IngressIdentities[id]; ok {
+		log.WithField(logfields.Identity, id).Debug("Removing ingress identity")
```
"Removing identity from ingress Map"?

ianvernon replied:
Replying to this comment in response to your review - I agree 100%, this might be confusing to end-users. Will fix such messages in my next upload :-)

ianvernon replied:
Done.

joestringer commented on pkg/policy/consumer.go:
```diff
@@ -258,37 +246,36 @@ func (c *Consumable) AllowConsumerAndReverseLocked(cache *ConsumableCache, id Nu
 		logfields.Identity + ".from": c.ID,
 		logfields.Identity + ".to":   id,
-	}).Warn("Allowed a consumer which can't be found in the reverse direction")
+	}).Warn("Allowed an ingress security identity which can't be found in the reverse direction")
```
Suggestion:
`Allowed a security identity on ingress, but couldn't [find/resolve] the identity for the reverse direction`

ianvernon replied:
Done.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -439,35 +440,34 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *policy.IdentityC
-			// created by a L3-L4 rule.
-			if _, ok := rulesRm[val.ID]; !ok {
-				rulesRm[val.ID] = policy.NewL4RuleContexts()
+			// Since we have removed an allowed ingress security identity, the
```
`Removed a security identity from ingress`?

ianvernon replied:
Done.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -439,35 +440,34 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *policy.IdentityC
-			// created by a L3-L4 rule.
-			if _, ok := rulesAdd[val.ID]; !ok {
-				rulesAdd[val.ID] = policy.NewL4RuleContexts()
+			// Since we have (re)added an ingress security identity, the L3 rule
```
`a security identity to ingress`?

ianvernon replied:
Done.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -478,12 +478,12 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *policy.IdentityC
+		"rulesAdd":                  rulesAdd,
+		"l4Rm":                      l4Rm,
+		"rulesRm":                   rulesRm,
+	}).Debug("New consumable with ingress security identities")
```
`Consumable regenerated` ? The `with ...` doesn't seem to add much here, the relevant fields are already included in the fields above.

ianvernon replied:
Done.

joestringer commented on pkg/policy/consumer.go:
```diff
@@ -216,40 +199,45 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
 			logfields.Identity: id,
 			"consumable":       logfields.Repr(c),
-		}).Debug("New consumer Identity for consumable")
+		}).Debug("New ingress security identity for consumable")
```
`Allowing security identity on ingress to consumable`?

ianvernon replied:
Done.

joestringer reviewed (commented):
LGTM overall, I think this makes the code a bit easier to follow.

That said, I think we should probably be a bit more careful about the usage of `Ingress identity` which suggests that an identity is defined in part by the direction of the traffic. IMHO Identities should be a representation of the labels, but should not be tied to policy or traffic flows---policy applies to an identity and applies to traffic flows in ingress/egress to/from an endpoint. For `Consumable.IngressIdentities` I think that this is fine, because it's describing the list of identities allowed on ingress to that consumable. However, some of the log messages are now referring to "Ingress identities" as though you could have two identities represented by the same number, with at least differing directions (ingress/egress) if not other differing attributes.

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -944,7 +944,7 @@ definitions:
         description: Build number of calculated policy in use
         type: integer
-      allowed-consumers:
+      allowed-ingress-identities:
```
Is this whole object marked in some way as "may change without notice"? It's leaking internal policy structures that I expect will change in subsequent Cilium versions.

ianvernon replied:
I'm not entirely sure what you mean by this -   the API is not yet stable since we have not officially released v1.0.

joestringer replied:
IMHO most of the fields in the endpoint here are implementation details which we should be free to change. I don't really think it should be formalised in the API, but if it is then I'm proposing that we mark it as an unstable portion of the API, so people shouldn't expect to rely on it.

ianvernon replied:
I agree about marking this portion of the API as unstable. However, I am of the belief that we need to have clear use of our vocabulary throughout all parts of the Cilium - it is not clear to a user of the API was a 'consumer' is, but 'identity' is used throughout.

joestringer replied:
Agree, 'consumer' isn't mentioned in the docs at all but 'identity' is described in the Concepts section.

Commit:
clarify addition of security identities for ingress policy

Rephrase some comments and log messages to specify that we are tracking security
policies for ingress, not "ingress security policies", which may be confusing.

Signed-off by: Ian Vernon <ian@cilium.io>

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -619,17 +618,17 @@ func (e *Endpoint) GetPolicyModel() *models.EndpointPolicy {
-	consumers := []int64{}
-	for _, v := range e.Consumable.Consumers {
-		consumers = append(consumers, int64(v.ID))
+	ingressIdentities := []int64{}
```
```
ingressIdentities := make([]int64, 0, len(e.Consumable.Consumers))
```

ianvernon replied:
Fixed.

Commit:
pkg/endpoint: create slice with capacity

Be more efficient in creation of slice of identities in GetPolicyModel, as we
know how many entries will be added to it.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -448,28 +449,27 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *policy.IdentityC
-		if val.DeletionMark {
-			val.DeletionMark = false
-			c.BanConsumerLocked(val.ID)
+	for val, ok := range c.Consumers {
```
this should not be `val, ok` it should be something like `identity, deny` and you can swap the if statement bellow

ianvernon replied:
This is addressed by a later commit in the PR.

aanm commented on pkg/policy/consumer.go:
```diff
@@ -65,7 +64,7 @@ func NewConsumable(id NumericIdentity, lbls *Identity, cache *ConsumableCache) *
 		Maps:         map[int]*policymap.PolicyMap{},
-		Consumers:    map[NumericIdentity]*Consumer{},
+		Consumers:    map[NumericIdentity]bool{},
 		ReverseRules: map[NumericIdentity]*Consumer{},
```
Why is Consumers a `map[NumericIdentity]bool,` but `ReverseRules` not a `map[NumericIdentity]bool` as well?

ianvernon replied:
This is addressed by a later commit in the PR.

aanm commented on pkg/policy/consumer.go:
```diff
@@ -211,8 +206,8 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
 // IngressIdentities map, and thus had to be added, false if it is already added.
 func (c *Consumable) AllowIngressIdentityLocked(cache *ConsumableCache, id NumericIdentity) bool {
-	isIdentityAllowed := c.isIdentityAllowed(id)
-	if isIdentityAllowed == false {
```
What happens if the identity exists but is not allowed? Wait, why was the function `isIdentityAllowed` removed? I think it's better to leave it, it's more simple to call the function without worrying about what's the meaning of the value/keys of a map

ianvernon replied:
> What happens if the identity exists but is not allowed? Wait, why was the function isIdentityAllowed removed? I think it's better to leave it, it's more simple to call the function without worrying about what's the meaning of the value/keys of a map

I disagree. Now it's just a check for existence in the map, which is provided by Golang. I don't see why we need to add a wrapper around something that Golang provides out of the box with just one line.

aanm replied:
I had to read th whole code. Makes sense

aanm commented on pkg/policy/consumer.go:
```diff
@@ -184,7 +184,11 @@ func (c *Consumable) addToMaps(id NumericIdentity) {
 
 func (c *Consumable) wasLastRule(id NumericIdentity) bool {
-	return c.ReverseRules[id] == false && c.IngressIdentities[id] == false
+	// A rule is the 'last rule' for an identity if it does not exist as a key
```
Make this comment as part of wasLastRule's documentation

ianvernon replied:
Fixed in latest upload.

Commit:
pkg/policy: add comment in wasLastRule to be outside of it

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme reviewed (approved):
Lookin' good

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -71,13 +71,13 @@ func (e *Endpoint) checkEgressAccess(owner Owner, dstLabels labels.LabelArray, o
-// allowConsumer must be called with global endpoint.Mutex held
-func (e *Endpoint) allowConsumer(owner Owner, id policy.NumericIdentity) bool {
+// allowIngressIdentity must be called with global endpoint.Mutex held
+func (e *Endpoint) allowIngressIdentity(owner Owner, id policy.NumericIdentity) bool {
```
This could probably become something more like `allowIdentityOnIngress()`, but I don't feel strongly enough that I think this should block the PR.

ianvernon replied:
Good idea. I'll change the name with my upcoming egress changes.

joestringer reviewed (approved):
LGTM.

rlenglet reviewed (approved):
LGTM. Thanks!
